rfc/0000-lock-file.md

```diff
@@ -0,0 +1,280 @@
+
+The distinction between `path` and non-`path` lint crate dependencies is unfortunate, but this is the least "evil" that was found.
+
+## Alternative. No special case for `path` dependencies
+
+In this case `path` dependencies appear in the `Marker.lock` as well. This would simplify the implementation for us and make the lock file behavior a bit more consistent. But is it worth it? The workspace's `Cargo.lock` file will contain an entry for the in-workspace crate that doesn't actually represent how that crate is built. That may be a potential footgun. I already imagine someone debugging their lint crate, jumping to definition into the `marker_api` in their IDE, reading the code that they think would be compiled, but when they run their lint the `marker_api` version that is actually used isn't the one that they see in their IDE.
+
+The advantage of this approach is consistency. Maybe we can make this work if `cargo marker` validates that the `marker_api` version (if present) in the linted workspace's `Cargo.lock` is equal to the version of `marker_api` in `Marker.lock`. But what if the lint crate uses something more than `marker_api`? We would need to validate the whole dependency tree of the lint crate in the workspace. And what if there are version differences? We should somehow synchronize `Marker.lock` with the workspace's `Cargo.lock`. That's not an easy task.
+
+## Alternative. Compile all lint crates in isolation within their workspaces.
+
+This means we do N distinct `cargo rustc --crate-type cdylib` invocations in the context of each crate's directory. For 3rd-party crates we would run this command in their directory in cargo home using the `Cargo.lock` file that they have there (the one that was published with the crate). This is arguably slower, and each lint crate will have its own patch version of `marker_api` compiled with it.
+
+When the lint crates are included from `crates.io` they are not guaranteed to have a `Cargo.lock` file, and, besides, that lock file isn't easily visible to the users. Tools like `cargo deny / cargo tree` aren't designed to reach out to that `Cargo.lock` file. Besides, no `cargo update` can be done for those lock files, so users are stuck with the versions of the dependencies that those 3rd-party crates have in their lock file.
+
+## Alternative. Not having `marker_api::lint_crate! {}`
+
+This macro adds the "dispatch" lint pass glue to the crate where it is defined and thus makes it possible to compile it in isolation via `cargo rustc --crate-type cdylib`. We could remove such requirement, and instead always compile lint crates with the `marker_rustc_driver` implementation where we could have special handling configured via an env var, for example. That special handling would implicitly add the "dispatch" lint pass to the crate's `lib.rs`.
+
+This won't be easy to implement, but we could look into that in the future.
```

Comment:
> Specifically, I'm thinking of how useful it would be to use extern crate marker_api with the same artifact that the driver was compiled with, instead of having it as a normal dependency. If the same artifact is used, we could remove all the ABI compatibility stuff and just call everything directly.

Actually, it's a bit a magic for me as to how linking to rustc private creates works, because Rust doesn't have a stable ABI, but it is still possible to slap an `extern crate rustc_driver;`. I wonder how this ABI works when `rustc_driver` uses its own `tracing` crate, and we use our own, for example. 

In any case, it works only with `#![feature(rustc_private)]`. I don't think it's possible to make `extern crate marker_api` work at all. If you know how to implement it, I'd be interested to hear

Reply:
It has been some time, since I read into this, but here is the basic breakdown how I remember it:

rustc compiles normal library crates to `.rlib` files. These store the types and functions defined in the crate and also define the ABI of them. This means that every crate using a specific `rlib` file will have ABI for every type defined in the `rlib` file.

The `extern crate XYZ` item can be used to declare, that this crate depends on a crate called `XYZ`. The interesting part is, that this way of declaring a dependency doesn't require the dependency to be specified inside `Cargo.toml`. It basically just says that some `rlib` file will be provided to the compiler/linker somehow. This allows the reuse of `rlib` files, and therefore allow ABI compatibility.

The simplest example is probably Rust's `std` crate. Every toolchain contains a precompiled version of the standard library. The prelude adds a `extern crate std` item and the compiler discovers the `rlib` file. This is the reason why you never need to compile the standard library locally.

Rustc provides `rlib` files for its crates like `rustc_hir` but restricts them to the `#![feature(rustc_private)]` feature. This is how ABI compatibility is ensured.

Note: I'm not 100% sure if it's the `rlib` file or another one, but I'm pretty sure that the general setup and idea is correct.

---

When I first read into this, there were two problems:
1. Marker didn't have any precompiled binaries. The user would always build Marker locally. Storing the rlib file and validating that its the same, the driver was compiled with etc sounded complicated.
2. The main reason why I decided against this, was that I wanted to allow users to check or compile lint crates with cargo, like any other crate. Having an `extern crate marker_api` requires that the `rlib` of `marker_api` is always made available to the compiler, and I don't know how to permanently add it to the `rlib`s that rustc discovers. If we figure this out, it might be possible to switch over to this model. We might even be able to redesign a part of the API since crates could bind directly to the driver.

Reply:
I've been thinking a tiny bit more about `rlib`s, as it would avoid a lot of boilerplate code and could make driver/adapter development easier. There are two things I considered:
1. Since we provide precompiled builds of the driver or build the driver locally, we should have access to the `marker_api.rlib` file used for the driver. We could try to inject it, either in the folders rustc already searches for `rlib` files or using `rustflags = ["--extern marker_api=path/to/libmarker_api.rlib"]` in the `.cargo/config.toml` file.
2. We could differentiate between normal cargo and cargo-marker builds.
    * During lint crate development, the user would use the `marker_api` version from the dependencies. For example, when running `cargo check` or `cargo clippy`.
    * When running `cargo marker` or `cargo maker test-setup` we control the compilation. We would rebuild the dylint and inject `marker_api.rlib` then, to make it ABI. This would be made even easier, by the change suggested in the RFC of combining all lint crates into one.

I'll think a bit more about this, but wanted to share the thought.